### ISSUE TITLE
Sugar var() only knows numeric

### ISSUE THREAD

**Issue by eddelbuettel** [RcppCore/Rcpp#223]
This fails to compile:

```{.R}
cppFunction('double myVar(IntegerVector x) { return(var(x)); }') 
```

**Comment by wush978** [comment]
Dear Dirk,

I study the related header and have some ideas of implementing a template specialization function `sugar::Var<INTSXP,NA,T> var( const VectorBase<INTSXP,NA,T>& t)` to resolve this issue.

Directly replace the `REALSXP` to `INTSXP` will produce the wrong result because the `Minus_Vector_Primitive<RTYPE,NA,T> mm( object, m) ;` will trim result to `int`. We need to rewrite the `get` method.

The simplest approach is to internally transform the `IntegerVector` to `NumericVector` and apply the original `var`. 

How do you think?

**Comment by eddelbuettel** [comment]
But an internal cast would create a copy and that is not ideal.

I think the best way, sadly, is to add four specialization as I did for `mean()` (cf #219 and #221).   It's one of those things I may get to one day when there is nothing more urgent on the plate.  

If you want to play with it, this (like a few of the other open issue tickets) is a nice one to create a contribution.  

And thanks also for switching to discussion first rather than pull request first ;-)

**Comment by wush978** [comment]
I'll take this. And thanks for your hint, I'll study your implementation of `mean()` as a good start.

Ya, discuss first does help!

**Comment by wush978** [comment]
In R, the `stats::var` drops the imaginary part, so I do the same thing in 0eb6b3a . 

If the unit test in 41fea1e is all what we need (test 4 different type and compare the result with `stats::var`), I will start to test the implementation 0eb6b3a on major machines and compilers.

**Comment by wush978** [comment]
Dear all,

I don't know if the test is adaquate or not, so I come here for helps. It would be great if any of you can spend 1 minute to review 41fea1e . It compare the result of `Rcpp::var` on 4 different types with the `stats::var`.

If the unit test is all what we need, could I send a PR after making the test passed on major machines and compilers?

**Comment by eddelbuettel** [comment]
Yes, that is the general idea. 

In practice most people only use `NumericVector` which is how we got by so far.  But it would be nice to complement it.  Not sure anybody ever uses `var(myLogicalVector)` ... but hey, nice to have.

**Comment by MattPD** [comment]
Question: Hasn't this commit per chance introduced a regression corresponding to the switch from formula (1) to formula (2) in the following:
http://davegiles.blogspot.com/2015/06/statistical-calculations-numerical.html ?

I'm referring to the following snippet (although `CPLXSXP` looks analogous):
https://github.com/jmp75/Rcpp/commit/03597298c8bac7ce77eb200bd31bb51f4597668e#diff-bfb36f3ad0e86b5efdcbca14acfdaba1L36

I'm wondering, was this intentional?
// The issue is numerical instability & potential negativity: https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance

**Comment by MattPD** [comment]
Hm, it seems like it does.
Here's a sample snippet demonstrating the issue:

```r
vars = function(x)
{
	xbar = mean(x)
	x2 = sum(x ^ 2)
	var1 = sum((x - xbar) ^ 2) / (n - 1)
	var2 = sum(x2 - n * xbar ^ 2) / (n - 1)
	c(var1, var2)
}

Rcpp::cppFunction('double var_rcpp(NumericVector x) { return var(x); }') 

set.seed(100)
n = 1e3
x = rnorm(n)

vars_result = vars(x)
var1 = vars_result[1]
var2 = vars_result[2]
var3 = var_rcpp(x)
discrepancy = var1 - var2
discrepancy_cpp = var1 - var3
cat("benign-data discrepancy, R = ", discrepancy, "\n")
cat("benign-data discrepancy, Rcpp = ", discrepancy_cpp, "\n")

x = x + 1e7
vars_result = vars(x)
var1 = vars_result[1]
var2 = vars_result[2]
var3 = var_rcpp(x)
discrepancy = var1 - var2
discrepancy_cpp = var1 - var3
cat("post-offset discrepancy, R = ", discrepancy, "\n")
cat("post-offset discrepancy, Rcpp = ", discrepancy_cpp, "\n")
```

Output:
```
benign-data discrepancy, R =  2.220446e-16 
benign-data discrepancy, Rcpp =  6.661338e-16 
post-offset discrepancy, R =  0.005054458 
post-offset discrepancy, Rcpp =  0.05310251
```

**Comment by eddelbuettel** [comment]
I am open for suggestions.  When I last looked at that I could not create cases with discrepancy.

On the margin, "do what R does but in C++" is a good policy.  We may want to be faster at times, but not generally give up precision.

**Comment by MattPD** [comment]
Right. Will look into comparison with R further (as in: right now I think restoring the previous formula is a strict improvement, but also want to investigate what -- if anything -- breaks if we stop there).

**Comment by eddelbuettel** [comment]
I am still a little woolly from all the traveling but already :+1: for re-raising this issue.  It needs work.

**Comment by wush978** [comment]
Dear @MattPD and @eddelbuettel ,

Thanks for your great suggestion of the formula. I am glad to help you fix it. I think we need to put a minimal unit test to address this issue and rewrite the implementation of the `var` function. Your demo is  looked great to be a unit test so I'll start from this ASAP.

Wush

**Comment by eddelbuettel** [comment]
:+1: 

We really do have several issues somewhat comingled:

- [ ] `var()`, `sd()`, `mean()`, ... only know numeric and int
- [ ] `var()`, `sd()`, `mean()`, ... use a one-pass algo
- [ ] `var()`, `sd()`, `mean()`, ... may produce incorrect results

Two and three are related, of course.  I tried to construct tests / counterexamples for three and failed, but then luckily @mattpd came along ...

**Comment by wush978** [comment]
Dear @MattPD ,

I think the current implementation has been the formula (2) in http://davegiles.blogspot.tw/2015/06/statistical-calculations-numerical.html , right? However, the unit test in https://github.com/RcppCore/Rcpp/compare/master...wush978:bugfix/var-numerical-instability?expand=1 still fails.

**Comment by MattPD** [comment]
Right.

Did further investigation.
I'm inclined to go with the stable formula.
I've looked into the R's implementation of `cov` and I'm not convinced the extra passes actually buy something in practice. Based on the numerical experiments so far (attached below, comments welcome): The stable algorithm doesn't seem to fail when R's doesn't -- and when it fails then R's does, too. That being said, I do like Thomas' idea of the introduced-error test, I think we may want to incorporate this together with examples by David E. Giles and/or John D. Cook (references either already given above or included in this comment; proposed unit tests to follow; @wush978 -- should we coordinate?).

I may be wrong, so I'm attaching my reasoning (so, feedback / course correction are definitely welcome; and I _do_ apologize about the length of this -- I feel it's not a bad idea to have this discussion documented somewhere in case I've missed something, let me know if there's a better venue for this; will split into multiple chunks so it's at least more manageable, hopefully).

**Comment by MattPD** [comment]
Input:

```r
# Idea, author: John D. Cook
# Idea, URL: http://www.johndcook.com/blog/2008/09/26/comparing-three-methods-of-computing-standard-deviation/
set.seed(100)
n = 1e6
x = runif(n)

report_discrepancies(x, "unif-benign")
report_discrepancies(x + 1e9, "unif-offset")
report_discrepancies(x + 1e12, "unif-unpleasant")
report_discrepancies(x + 1e16, "unif-malign")
```

Output:
```
> report_discrepancies(x, "unif-benign")
var_R: 0.08321337 
unif-benign ; var_stable   = 0.08321337 	 discrepancy =  0 
unif-benign ; var_unstable = 0.08321337 	 discrepancy =  -4.163336e-17 
unif-benign ; var_rcpp     = 0.08321337 	 discrepancy =  7.771561e-16 

> report_discrepancies(x + 1e9, "unif-offset")
var_R: 0.08321337 
unif-offset ; var_stable   = 0.08321337 	 discrepancy =  0 
unif-offset ; var_unstable = 134.2179 	 discrepancy =  -134.1346 
unif-offset ; var_rcpp     = -3087.011 	 discrepancy =  3087.094 

> report_discrepancies(x + 1e12, "unif-unpleasant")
var_R: 0.08321338 
unif-unpleasant ; var_stable   = 0.08321338 	 discrepancy =  0 
unif-unpleasant ; var_unstable = 140737629 	 discrepancy =  -140737629 
unif-unpleasant ; var_rcpp     = -9.835871e+12 	 discrepancy =  9.835871e+12 

> report_discrepancies(x + 1e16, "unif-malign")
var_R: 0 
unif-malign ; var_stable   = 0 	 discrepancy =  0 
unif-malign ; var_unstable = -3.589002e+17 	 discrepancy =  3.589002e+17 
unif-malign ; var_rcpp     = -1.702131e+20 	 discrepancy =  1.702131e+20 
```

_Conclusion:_ no extra discrepancy introduced by using `stable` instead of `R`; remaining formulas not acceptable

**Comment by MattPD** [comment]
Input:

```r
# Idea, author: Thomas Lumley
# Idea, URL: http://davegiles.blogspot.com/2015/06/statistical-calculations-numerical.html?showComment=1433731907639#c7065245445825854941
set.seed(100)
x = rnorm(n)
set.seed(100)
y = rnorm(n) + 1e10

compute_introduced_error = function(x_var, y_var)
{
	introduced_error = x_var - y_var
	introduced_error
}

report_introduced_errors = function(x, y)
{
	x_vars_result = vars(x)
	x_var_R = var(x)
	x_var_stable = x_vars_result[1]
	x_var_unstable = x_vars_result[2]
	x_var_rcpp = var_rcpp_impl(x)
	
	y_vars_result = vars(y)
	y_var_R = var(y)
	y_var_stable = y_vars_result[1]
	y_var_unstable = y_vars_result[2]
	y_var_rcpp = var_rcpp_impl(y)
	
	cat("var_R_introduced_error:", compute_introduced_error(x_var_R, y_var_R), "\n")
	cat("var_stable_introduced_error:", compute_introduced_error(x_var_stable, y_var_stable), "\n")
	cat("var_unstable_introduced_error:", compute_introduced_error(x_var_unstable, y_var_unstable), "\n")
	cat("var_rcpp_introduced_error:", compute_introduced_error(x_var_rcpp, y_var_rcpp), "\n")
}

report_introduced_errors(x, y)
```

Output:
```
var_R_introduced_error: 5.774802e-10 
var_stable_introduced_error: 5.774803e-10 
var_unstable_introduced_error: 0.9982819 
var_rcpp_introduced_error: -412316.3 
```

_Conclusion_: no extra discrepancy introduced by using stable instead of R; remaining formulas not acceptable (unchanged).

**Comment by MattPD** [comment]
@wush978: Interesting, looking into it!

**Comment by MattPD** [comment]
@wush978: nice, you have found a good, new unit test!  :thumbsup:

Input:
```r
report_discrepancies(x = (1:10) + 1e10, data_case = "unit-test-wush978")
```

Output:
```
var_R: 9.166667 
unit-test-candidate ; var_stable   = 8.250008e-05 	 discrepancy =  9.166584 
unit-test-candidate ; var_unstable = -9.99991e+19 	 discrepancy =  9.99991e+19 
unit-test-candidate ; var_rcpp     = 0 	 discrepancy =  9.166667 
```

**Comment by MattPD** [comment]
@wush978: LOL, I had a bug in my test suite, NVM :blush:
I think I know how to fix this, give me a moment (or a few).

**Comment by MattPD** [comment]
This should do the trick: https://github.com/RcppCore/Rcpp/pull/320

```r
> report_discrepancies(x = (1:10) + 1e10, data_case = "unit-test-wush978")
var_R: 9.166667 
unit-test-wush978 ; var_stable   = 9.166667 	 discrepancy =  0 
unit-test-wush978 ; var_unstable = 0 	 discrepancy =  9.166667 
unit-test-wush978 ; var_rcpp     = 9.166667 	 discrepancy =  0 
```

I think we may want to also add more unit tests for this -- starting from the ones by @wush978.

Thanks again for that one, @wush978, since it's allowed me spot a missing length calculation in the `vars` function I've previously used for testing -- a mark of a truly good unit test :-)

Other than that, there's a similar issue with the complex specialization (as @eddelbuettel noticed): https://github.com/RcppCore/Rcpp/blob/master/inst/include/Rcpp/sugar/functions/var.h#L47

Currently looking for examples / ways to test, I don't think R's `var` function is an appropriate test case here: http://r.789695.n4.nabble.com/Calculate-variance-covariance-with-complex-numbers-td1693542.html

**Comment by MattPD** [comment]
Right. Found a good test case, exactly as in the aforementioned comment:
In accordance with its definition, the variance of {5+2i, 4+3i, 3+4i, 2+5i, 1+6i} is 5.

// This also agrees with, say, Wolfram Alpha: http://www.wolframalpha.com/share/clip?f=d41d8cd98f00b204e9800998ecf8427etfc6d73787

So this is what I'm going for in the implementation.

On a side note, R isn't implementing this at all (again, as previously mentioned), only the real case, hence the warnings about coercions we've been having in the our testcases (there was nothing to test against, since the implementation for complex numbers was missing, so the function call was resolved to the regular numeric type instead).

**Comment by coatless** [comment]
This issue was addressed in PR #320 

It's very good it was switched from a one pass method to a two pass. I actually used this as a basis for triggering an overflow.

http://stat385.thecoatlessprofessor.com/assets/lectures/lec6/lec6_numerical_stability_errors_and_debugging_techniques.pdf#page=14

Wiki also has a great overview of variance algorithms:

https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance

And, of course, a nice paper (albeit dated):

[Algorithms for Computing the Sample Variance: Analysis and Recommendations](http://www.jstor.org/stable/2683386?seq=1#page_scan_tab_contents)

Please close. (Tagging #506)

**Comment by eddelbuettel** [comment]
Hah. I once tried to trigger an overflow and basically "failed" as I thought I couldn't come up with a drastic enough example.
